core/src/main/scala/org/apache/spark/util/random/SamplingUtils.scala

```diff
@@ -52,16 +52,20 @@ private[spark] object SamplingUtils {
       (trimReservoir, i)
     } else {
       // If input size > k, continue the sampling process.
+      var l = i.toLong
       val rand = new XORShiftRandom(seed)
       while (input.hasNext) {
         val item = input.next()
-        val replacementIndex = rand.nextInt(i)
+        val replacementIndex = l < Int.MaxValue match {
+          case true => rand.nextInt(l.toInt)
+          case false => rand.nextInt()
```

Comment:
Ah, this won't be valid right? you're not choosing an element uniformly at random here. I know there's no `nextLong(Long)` but you can use (l * rand.nextDouble()).toLong